barretenberg/cpp/src/barretenberg/commitment_schemes/commitment_key.hpp

```diff
@@ -81,6 +82,65 @@ template <class Curve> class CommitmentKey {
         return scalar_multiplication::pippenger_unsafe<Curve>(
             const_cast<Fr*>(polynomial.data()), srs->get_monomial_points(), degree, pippenger_runtime_state);
     };
+
+    /**
+     * @brief Efficiently commit to a sparse polynomial
+     * @details Iterate through the {point, scalar} pairs that define the inputs to the commitment MSM, maintain (copy)
+     * only those for which the scalar is nonzero, then perform the MSM on the reduced inputs.
+     * @warning Method makes a copy of all {point, scalar} pairs that comprise the reduced input. Will not be efficient
+     * in terms of memory or computation for polynomials beyond a certain sparseness threshold.
+     *
+     * @param polynomial
+     * @return Commitment
+     */
+    Commitment commit_sparse(std::span<const Fr> polynomial)
+    {
+        BB_OP_COUNT_TIME();
+        const size_t degree = polynomial.size();
+        ASSERT(degree <= srs->get_monomial_size());
+
+        // Extract the precomputed point table (contains raw SRS points at even indices and the corresponding
+        // endomorphism point (\beta*x, -y) at odd indices).
+        G1* point_table = srs->get_monomial_points();
+
+        // Define structures needed to multithread the extraction of non-zero inputs
+        const size_t num_threads = degree >= get_num_cpus_pow2() ? get_num_cpus_pow2() : 1;
+        const size_t block_size = degree / num_threads;
+        std::vector<std::vector<Fr>> thread_scalars(num_threads);
+        std::vector<std::vector<G1>> thread_points(num_threads);
+
+        // Loop over all polynomial coefficients and keep {point, scalar} pairs for which scalar != 0
+        parallel_for(num_threads, [&](size_t thread_idx) {
+            const size_t start = thread_idx * block_size;
+            const size_t end = (thread_idx + 1) * block_size;
+
+            for (size_t idx = start; idx < end; ++idx) {
+
+                const Fr& scalar = polynomial[idx];
+
+                if (!scalar.is_zero()) {
+                    thread_scalars[thread_idx].emplace_back(scalar);
+                    // Save both the raw srs point and the precomputed endomorphism point from the point table
+                    const G1& point = point_table[idx * 2];
+                    const G1& endo_point = point_table[idx * 2 + 1];
+                    thread_points[thread_idx].emplace_back(point);
+                    thread_points[thread_idx].emplace_back(endo_point);
+                }
+            }
+        });
+
+        // Reconstruct the full input to the pippenger from the individual threads
+        std::vector<Fr> scalars;
```

Comment:
Probably worth reserving some space in these just in case / since it's easy.

Reply:
added